File: backend/internal/models/movie.go
package models

// MovieResponse represents the API response for a movie with spoiler details
type MovieResponse struct {
	Title    string   `json:"title"`
	Year     string   `json:"year"`
	Poster   string   `json:"poster"`
	Backdrop string   `json:"backdrop"`
	Rating   float64  `json:"rating"`
	Genres   []string `json:"genres"`
	Overview string   `json:"overview"`
	Spoiler  string   `json:"spoiler"`
}

// TMDBSearchResult represents the TMDB API search response
type TMDBSearchResult struct {
	Results []TMDBMovie `json:"results"`
}

// TMDBMovie represents a single movie from TMDB API
type TMDBMovie struct {
	ID            int    `json:"id"`
	Title         string `json:"title"`
	ReleaseDate   string `json:"release_date"`
	PosterPath    string `json:"poster_path"`
	BackdropPath  string `json:"backdrop_path"`
	VoteAverage   float64 `json:"vote_average"`
	Overview      string `json:"overview"`
	GenreIDs      []int  `json:"genre_ids"`
}

// TMDBGenreResponse represents the genres from TMDB API
type TMDBGenreResponse struct {
	Genres []TMDBGenre `json:"genres"`
}

// TMDBGenre represents a genre from TMDB API
type TMDBGenre struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

// GeminiRequest represents the request structure for Gemini API
type GeminiRequest struct {
	Contents []GeminiContent `json:"contents"`
}

// GeminiContent represents content for Gemini API
type GeminiContent struct {
	Parts []GeminiPart `json:"parts"`
}

// GeminiPart represents a text part for Gemini API
type GeminiPart struct {
	Text string `json:"text"`
}

// GeminiResponse represents the response from Gemini API
type GeminiResponse struct {
	Candidates []GeminiCandidate `json:"candidates"`
}

// GeminiCandidate represents a candidate response from Gemini API
type GeminiCandidate struct {
	Content GeminiContent `json:"content"`
}

// ErrorResponse represents an error response
type ErrorResponse struct {
	Error string `json:"error"`
}

File: frontend/types/movie.ts
/**
 * TypeScript interfaces matching the backend API responses.
 * Keep in sync with backend/internal/models/movie.go
 */

/** Single movie — returned by /api/movie and as items in lists */
export interface Movie {
  title: string;
  year: string;
  poster: string;
  backdrop: string;
  rating: number;
  genres: string[];
  overview: string;
  spoiler?: string; // Only present when fetched with spoiler
  runtime?: number; // Runtime in minutes (from TMDB detail endpoint)
}

/** Parsed character fate from Gemini spoiler text */
export interface CharacterFate {
  name: string;
  actor: string;
  status: "ALIVE" | "DEAD" | "UNKNOWN";
  summary: string;
}

/** Parsed key moment from Gemini spoiler text */
export interface KeyMoment {
  title: string;
  description: string;
}

/** Parsed story section from Gemini spoiler text */
export interface StorySection {
  id: string;
  title: string;
  icon: string;
  content: string;
}

/** Parsed interpretation section */
export interface InterpretationSection {
  title: string;
  content: string;
}

/** Response shape from /api/movies and /api/search */
export interface MoviesResponse {
  movies: Movie[];
  count: number;
  year?: string;
  query?: string;
  page?: number;
  total_pages?: number;
}

/** Response shape from /api/trending */
export interface TrendingResponse {
  movies: Movie[];
  count: number;
}

/** Backend error response */
export interface ApiError {
  error: string;
}
